packages/dds/tree/src/simple-tree/proxies.ts

```diff
@@ -193,15 +194,25 @@ export function createObjectProxy<TSchema extends FlexObjectNodeSchema>(
 
 			return true;
 		},
-		has: (target, key) => {
-			return (
-				schema.objectNodeFields.has(key as FieldKey) ||
-				(allowAdditionalProperties ? Reflect.has(target, key) : false)
-			);
+		// The "has" trap is called when a user checks for a property with "in" (e.g. `if ("z" in point) {...}`).
+		has: (target, key: FieldKey) => {
+			const fieldSchema = schema.objectNodeFields.get(key);
+			if (fieldSchema !== undefined) {
+				if (fieldSchema.kind.multiplicity === Multiplicity.Optional) {
+					const field = getFlexNode(proxy).tryGetField(key);
+					if (field === undefined) {
+						return false; // Don't include the keys of optional undefined fields
+					}
+				}
+				return true;
+			}
+
+			return allowAdditionalProperties ? Reflect.has(target, key) : false;
 		},
+		// The "ownKeys" trap is called by Reflect.ownKeys(), Object.keys() and for-in loops.
 		ownKeys: (target) => {
 			return [
-				...schema.objectNodeFields.keys(),
+				...filterIterable(schema.objectNodeFields.keys(), (k) => k in proxy),
 				...(allowAdditionalProperties ? Reflect.ownKeys(target) : []),
```

Comment:
In "has" we are allow empty optional fields to block access to properties with the same name on the target. THis seems like a reasonable choice.

Here we are assuming that the target doesn't not have any keys which exist in the schema. Also a reasonable choice.

We don't document either of them and they don't match.

Reply:
Good observation. I'm not sure what our behavior is supposed to be for "field shadowing", but I agree they should be consistent. We need to decide:

1. Do fields shadow "extra properties" or do "extra properties" shadow fields?
2. When the shadower is undefined, do we allow reads to fall back to the shadowed?

Do you have an opinion one way or the other? @DLehenbauer, do you? 

My intuition is to say that extra properties should shadow fields. Since we don't provide any "built-in" properties on objects, the only way a shadowing occurs is if a user does it themselves. For example, they might add a field name "foo" and also put a getter method on their class-based schema object named "foo()". I'd expect that to mean they are writing a getter that reads the foo field under and does some kind of validation or something before returning it... in which case, they would want "foo()" to shadow "foo". 

If extra properties shadow fields, then we know what to do when the property is absent - we fall back to the field. But what if the property is present, but set to `undefined`? Do we return undefined, or do we fall back? If we fall back, then there is no distinction between "undefined present" and "undefined absent" and that's consistent with how our fields work in general. But, then it's not consistent with JS in general, where you will fall back to the prototype of an object if a property is missing, but not if that property is explicitly set to `undefined`. Ugh.